request-deferrer/service-worker.js

```diff
@@ -0,0 +1,183 @@
+    return new Response();
+  });
+});
+
+// Start the service worker.
+worker.init();
+
+// By using Mozilla's localforage db wrapper, we can count on
+// a fast setup for a versatile key-value database. We use
+// it to store queue of deferred requests.
+
+// Enqueue consists on adding a request to the list. Due to the
+// limitations of IndexedDB, Request and Response objects can not
+// be saved so we need an alternative representations. This is
+// why we call to `serialize()`.`
+function enqueue(request) {
+  return serialize(request).then(function(serialized) {
+    localforage.getItem('queue').then(function(queue) {
+      /* eslint no-param-reassign: 0 */
+      queue = queue || [];
+      queue.push(serialized);
+      return localforage.setItem('queue', queue).then(function() {
+        console.log(serialized.method, serialized.url, 'enqueued!');
+      });
+    });
+  });
+}
+
+// Flush is a little more complicated. It consists into get
+// the elements of the queue in order and send each one,
+// keeping track of not yet sent request. Before sending a request
+// we need to recreate it from the alternative representation
+// stored in IndexedDB.
+function flushQueue() {
+  // Get the queue
+  return localforage.getItem('queue').then(function(queue) {
+    /* eslint no-param-reassign: 0 */
+    queue = queue || [];
+
+    // If empty, nothing to do!
+    if (!queue.length) {
+      return Promise.resolve();
+    }
+
+    // Else, send the requests in order...
+    console.log('Sending ', queue.length, ' requests...');
+    return sendInOrder(queue).then(function() {
+      // **Requires error handling**. Actually, this is assuming all the requests
+      // in queue are a success when reaching the Network. So it should empty the
+      // queue step by step, only popping from the queue if the request completes
+      // with success.
+      return localforage.setItem('queue', []);
+    });
+  });
+}
+
+// Send the requests inside the queue in order. Waiting for the current before
+// sending the next one.
+function sendInOrder(requests) {
+  // The `reduce()` chains one promise per serialized request, not allowing to
+  // progress to the next one until completing the current.
+  var sending = requests.reduce(function(prevPromise, serialized) {
+    console.log('Sending', serialized.method, serialized.url);
+    return prevPromise.then(function() {
+      return deserialize(serialized).then(function(request) {
+        return fetch(request);
+      });
+    });
+  }, Promise.resolve());
+  return sending;
+}
+
+// Serialize is a little bit convolved due to headers is
+// a [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)
+// and not a simple object.
+function serialize(request) {
+  var headers = {};
+  // `for(... of ...)` is ES6 notation but current browsers supporting SW, support this
+  // notation as well and this is the only way of retrieving all the headers.
```

Comment:
Talked with @delapuente on IRC, `forEach` isn't currently available for Headers.
Headers is not a Map though, so the comment needs to be fixed.